linker/resolve.go

```diff
@@ -152,20 +152,24 @@ func descriptorTypeWithArticle(d protoreflect.Descriptor) string {
 	}
 }
 
-func (r *result) resolveReferences(handler *reporter.Handler, s *Symbols) error {
-	// first create the full descriptor hierarchy
+func (r *result) createDescendants() {
 	fd := r.FileDescriptorProto()
+	pool := &allocPool{}
+	pool.countElements(fd)
```

Comment:
I think having a constructor here would be a nicer api: 
```go
pool := newPool(fd)
```
The other `count...` methods look like helpers for `countElements`.

Reply:
Done. Still have the count fields, to accumulate counts as this descends recursively through the hierarchy. But it does now eagerly allocate the slices instead of doing so lazily.